torch/_ops.py

```diff
@@ -409,6 +430,93 @@ def key_extractor(tensors, key_mask):
 _mode_stack_per_key: Dict[torch._C.DispatchKey, List] = {}
 
 
+class ModeSlot(enum.IntEnum):
+    PROXY_MODE_SLOT = 0
+    FUNCTIONAL_MODE_SLOT = 1
+
+
+_mode_stack_per_key[torch._C.DispatchKey.PreDispatch] = [None, None]
+
+
+def _get_pre_dispatch_slot(mode_key):
+    assert mode_key in (
+        torch._C._TorchDispatchModeKey.PROXY,
+        torch._C._TorchDispatchModeKey.FUNCTIONAL,
+    )
+    if mode_key == torch._C._TorchDispatchModeKey.PROXY:
+        return ModeSlot.PROXY_MODE_SLOT
+    return ModeSlot.FUNCTIONAL_MODE_SLOT
+
+
+def unset_mode_pre_dispatch(mode_key):
+    current_mode_stack_pre_dispatch = mode_stack_per_key()[
+        torch._C.DispatchKey.PreDispatch
+    ]
+    assert mode_key in (
+        torch._C._TorchDispatchModeKey.PROXY,
+        torch._C._TorchDispatchModeKey.FUNCTIONAL,
+    )
+    slot = _get_pre_dispatch_slot(mode_key)
+    current_mode = current_mode_stack_pre_dispatch[slot]
+    mode_stack_per_key()[torch._C.DispatchKey.PreDispatch][slot] = None
+    return current_mode
+
+
+def _set_mode_pre_dispatch(mode):
+    assert hasattr(mode, "_mode_key") and mode._mode_key in (
+        torch._C._TorchDispatchModeKey.PROXY,
+        torch._C._TorchDispatchModeKey.FUNCTIONAL,
+    )
+    mode_key = mode._mode_key
+    slot = _get_pre_dispatch_slot(mode_key)
+    current_mode = mode_stack_per_key()[torch._C.DispatchKey.PreDispatch][slot]
+    assert current_mode is None
+    mode_stack_per_key()[torch._C.DispatchKey.PreDispatch][slot] = mode
+
+
+def _pop_mode_from_pre_dispatch():
+    for i in range(1, -1, -1):
```

Comment:
nit: The for loop here makes me think that this is "generic", but we're really just encoding the fact that there are exactly two modes in the loop iteration. I would just if/else on whether the functional / proxy modes are set.